src/qibolab/instruments/zhinst/executor.py

```diff
@@ -0,0 +1,781 @@
+
+        self.experiment_flow(qubits, couplers, sequence, options)
+
+        self.run_exp()
+
+        # Get the results back
+        results = {}
+        for qubit in qubits.values():
+            q = qubit.name  # pylint: disable=C0103
+            for i, ropulse in enumerate(self.sequence[measure_channel_name(qubit)]):
+                data = np.array(self.results.get_data(f"sequence{q}_{i}"))
+                if options.acquisition_type is AcquisitionType.DISCRIMINATION:
+                    data = (
+                        np.ones(data.shape) - data.real
+                    )  # Probability inversion patch
+                serial = ropulse.pulse.serial
+                qubit = ropulse.pulse.qubit
+                results[serial] = results[qubit] = options.results_type(data)
+
+        return results
+
+    def sequence_zh(
+        self, sequence: PulseSequence, qubits: dict[str, Qubit]
+    ) -> dict[str, list[ZhPulse]]:
+        """Convert Qibo sequence to a sequence where all pulses are replaced
+        with ZhPulse instances.
+
+        The resulting object is a dictionary mapping from channel name
+        to corresponding sequence of ZhPulse instances
+        """
+        # Define and assign the sequence
+        zhsequence = defaultdict(list)
+
+        # Fill the sequences with pulses according to their lines in temporal order
+        for pulse in sequence:
+            if pulse.type == PulseType.READOUT:
+                ch = measure_channel_name(qubits[pulse.qubit])
+            else:
+                ch = pulse.channel
+            zhsequence[ch].append(ZhPulse(pulse))
+
+        if self.processed_sweeps:
+            for ch, zhpulses in zhsequence.items():
+                for zhpulse in zhpulses:
+                    for param, sweep in self.processed_sweeps.sweeps_for_pulse(
+                        zhpulse.pulse
+                    ):
+                        zhpulse.add_sweeper(param, sweep)
+
+        return zhsequence
+
+    def create_exp(self, qubits, options):
+        """Zurich experiment initialization using their Experiment class."""
+
+        # Setting experiment signal lines
+        signals = [lo.ExperimentSignal(name) for name in self.signal_map.keys()]
+
+        exp = lo.Experiment(
+            uid="Sequence",
+            signals=signals,
+        )
+
+        if self.acquisition_type:
+            acquisition_type = self.acquisition_type
+            self.acquisition_type = None
+        else:
+            acquisition_type = ACQUISITION_TYPE[options.acquisition_type]
+        averaging_mode = AVERAGING_MODE[options.averaging_mode]
+        exp_options = replace(
+            options, acquisition_type=acquisition_type, averaging_mode=averaging_mode
+        )
+
+        # Near Time recursion loop or directly to Real Time recursion loop
+        if self.nt_sweeps:
+            self.sweep_recursion_nt(qubits, exp_options, exp)
+        else:
+            self.define_exp(qubits, exp_options, exp)
```

Comment:
This is repeated below:
https://github.com/qiboteam/qibolab/blob/c982befc9865f35d86f046c9cabb31c8404c171c/src/qibolab/instruments/zhinst/executor.py#L775-L778

I would suggest to deduplicate, just by always calling the recursion, and checking at the beginning (instead of the end), if you have to stop or continue.